### PR TITLE
[JSC] add support for `RegExp.escape`

### PR DESCRIPTION
**Pull request by dcrousso**
#### 990058dd93878188d65e53e02eda4b64feb69c1d
<pre>
[JSC] add support for `RegExp.escape`
<a href="https://bugs.webkit.org/show_bug.cgi?id=272622">https://bugs.webkit.org/show_bug.cgi?id=272622</a>

Reviewed by Justin Michaud and Ross Kirsling.

This will help developers construct regular expressions that contain special characters without having to manually escape/replace them (e.g. URLs often contain a `.` so making a regular expression to match would require that all of them be replaced with `\.` instead).

Spec: &lt;<a href="https://tc39.es/proposal-regex-escaping/">https://tc39.es/proposal-regex-escaping/</a>&gt;
Proposal: &lt;<a href="https://github.com/tc39/proposal-regex-escaping">https://github.com/tc39/proposal-regex-escaping</a>&gt;

* Source/JavaScriptCore/runtime/RegExpConstructor.cpp:
(JSC::RegExpConstructor::finishCreation):
(JSC::regExpConstructorEscape): Added.

* Source/JavaScriptCore/runtime/OptionsList.h:
Add a feature flag for this.

* JSTests/stress/regexp-escape.js: Added.

Canonical link: <a href="https://commits.webkit.org/277793@main">https://commits.webkit.org/277793@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/5a11907fb9cc6443c0c4d7b0bd5fb76bed20bf7b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48591 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27802 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51552 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51278 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44656 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33739 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25331 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39736 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49173 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25467 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41922 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20832 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22944 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43096 "Found 1 new test failure: fast/css/viewport-height-border.html (failure)") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6647 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41890 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44883 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43583 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53186 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/48082 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23638 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19936 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47029 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24903 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42127 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45956 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10711 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25708 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55576 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24625 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11430 "Passed tests") |

### CODE REVIEW

**Comment by dcrousso** [comment]
this is not at stage 3 yet, but it's quite close so i figured i'd get the jump on it :)

as of today, the latest (unmerged) changes are actually at <[https://bakkot.github.io/proposal-regex-escaping/](https://bakkot.github.io/proposal-regex-escaping/)>

File: Source/JavaScriptCore/runtime/JSGlobalObject.cpp [hunk 1]
```diff
@@ -1162,7 +1162,7 @@ void JSGlobalObject::init(VM& vm)
     ShadowRealmConstructor* shadowRealmConstructor = ShadowRealmConstructor::create(vm, ShadowRealmConstructor::createStructure(vm, this, m_functionPrototype.get()), m_shadowRealmPrototype.get());
     m_shadowRealmConstructor.set(vm, this, shadowRealmConstructor);
 
-    RegExpConstructor* regExpConstructor = RegExpConstructor::create(vm, RegExpConstructor::createStructure(vm, this, m_functionPrototype.get()), m_regExpPrototype.get());
+    RegExpConstructor* regExpConstructor = RegExpConstructor::create(vm, this, RegExpConstructor::createStructure(vm, this, m_functionPrototype.get()), m_regExpPrototype.get());
```
Comment (rkirsling) [inline Source/JavaScriptCore/runtime/JSGlobalObject.cpp]: What's the purpose of this change? `RegExpConstructor` and `ErrorConstructor` seem very similar here, yet #27217 doesn't do this. 🤔

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp [hunk 2]
```diff
@@ -83,18 +86,96 @@ RegExpConstructor::RegExpConstructor(VM& vm, Structure* structure)
...
     putDirectNonIndexAccessorWithoutTransition(vm, vm.propertyNames->speciesSymbol, speciesGetterSetter, PropertyAttribute::Accessor | PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum);
 }
 
+JSC_DEFINE_HOST_FUNCTION(regExpConstructorEscape, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue argValue = callFrame->argument(0);
+    if (!argValue.isString())
```
Comment (rkirsling) [inline Source/JavaScriptCore/runtime/RegExpConstructor.cpp]: You could probably make this `UNLIKELY` if you want.

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp [hunk 3]
```diff
@@ -83,18 +86,96 @@ RegExpConstructor::RegExpConstructor(VM& vm, Structure* structure)
...
+        return throwVMTypeError(globalObject, scope, "RegExp.escape requires a string"_s);
+
+    JSString* argString = asString(argValue);
+    ASSERT(argString);
```
Comment (rkirsling) [inline Source/JavaScriptCore/runtime/RegExpConstructor.cpp]: This ASSERT seems superfluous; I think checking `v.isString()` and then doing `asString(v)->value(globalObject)` is a common pattern in JSC.

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp [hunk 4]
```diff
@@ -83,18 +86,96 @@ RegExpConstructor::RegExpConstructor(VM& vm, Structure* structure)
...
+            codePoint = string.characters8()[i++];
+        else
+            U16_NEXT(string.characters16(), i, string.length(), codePoint);
+
+        if (builder.isEmpty() && isASCIIAlphanumeric(codePoint)) {
+            builder.append('\\', 'x', toStringWithRadix(codePoint, 16));
+            continue;
+        }
+
+        if (isASCII(codePoint) && strchr("^$\\.*+?()[]{}|/", codePoint)) {
```
Comment (rkirsling) [inline Source/JavaScriptCore/runtime/RegExpConstructor.cpp]: `strchr` seems rather unusual; I would expect `.contains` here.

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp [hunk 5]
```diff
@@ -83,18 +86,96 @@ RegExpConstructor::RegExpConstructor(VM& vm, Structure* structure)
...
+            builder.append('\\', 'f');
+            continue;
+        case '\r':
+            builder.append('\\', 'r');
+            continue;
+        default:
+            break;
+        }
+
+        if ((isASCII(codePoint) && strchr(",-=<>#&!%:;@~'`\"", codePoint)) || isStrWhiteSpace(codePoint) || U16_IS_SURROGATE(codePoint)) {
```
Comment (rkirsling) [inline Source/JavaScriptCore/runtime/RegExpConstructor.cpp]: Ditto.

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp [hunk 6]
```diff
@@ -83,18 +86,96 @@ RegExpConstructor::RegExpConstructor(VM& vm, Structure* structure)
...
+            if (isLatin1(codePoint))
+                builder.append('\\', 'x', pad('0', 2, toStringWithRadix(codePoint, 16)));
+            else if (U_IS_BMP(codePoint))
```
Comment (rkirsling) [inline Source/JavaScriptCore/runtime/RegExpConstructor.cpp:160]: Here and below, I wonder if it would be more consistent to say `!U16_IS_SURROGATE`?

File: Source/JavaScriptCore/runtime/JSGlobalObject.cpp (hunk 1 above)
Comment (dcrousso) [reply Source/JavaScriptCore/runtime/JSGlobalObject.cpp]: yeah i realized i could do it a different way after making this change

im not sure which is better/preferred, as ive seen it be done both ways (and have personally done it this way at least once).  im pretty sure the `JSGlobalObject` used to create the `RegExpPrototype` is the same, but i figured it couldn't hurt to pass it along explicitly just in case :)

let me know if it's better to not do this

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp (hunk 4 above)
Comment (dcrousso) [reply Source/JavaScriptCore/runtime/RegExpConstructor.cpp]: this is using a raw string though, not a `String`.  do we have a `static` version of `contains` that takes a `const char*`?  or are you suggesting that i create a `String`/`StringView`?

FWIW ive seen this done in a few other places (e.g. `JSC::Yarr::Parser`, `Inspector::ContentSearchUtilities`, etc.)

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp (hunk 6 above)
Comment (dcrousso) [reply Source/JavaScriptCore/runtime/RegExpConstructor.cpp:160]: is there a specific reason for using `U16_IS_SURROGATE` instead of `U_IS_BMP`?  i was trying to follow what `stringFromCodePoint` does.

also technically i think all we really care about is `codePoint <= 0xFFFF` according to the spec <[https://tc39.es/ecma262/#sec-unicodeescape](https://tc39.es/ecma262/#sec-unicodeescape)>

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp (hunk 4 above)
Comment (rkirsling) [reply Source/JavaScriptCore/runtime/RegExpConstructor.cpp]: I was suggesting that it wouldn't just be a raw string, yeah—I think `AtomString` (i.e. `_s`) would be appropriate?
That said, I'd just been grepping in `runtime/`, so I missed the two cases you mentioned, which are indeed super similar.

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp (hunk 6 above)
Comment (rkirsling) [reply Source/JavaScriptCore/runtime/RegExpConstructor.cpp:160]: What I mean is that you're in the implicit "else" after [step 5a](https://bakkot.github.io/proposal-regex-escaping/#sec-encode), so having ruled out the Latin1 case, it would seem most natural to describe the follow-up question as "surrogate pair or not surrogate pair?".

(I assume that the code should work correctly either way, so it should just be a matter of readability.)

**Review by justinmichaud — APPROVED** [review]
LGTM too, once @rkirsling's feedback is addressed.

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp (hunk 4 above)
Comment (dcrousso) [reply Source/JavaScriptCore/runtime/RegExpConstructor.cpp]: so based on that is this fine to leave as-is or would you rather i change it to use an `AtomString` instead?  or maybe something like `charactersContain` instead?

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp (hunk 6 above)
Comment (dcrousso) [reply Source/JavaScriptCore/runtime/RegExpConstructor.cpp:160]: i dont really know enough about the intended use to comment so i'll trust your knowledge about it, but the spec specifically says to check for `<= 0xFFFF` ([step 2 of `UTF16EncodeCodePoint`](https://tc39.es/ecma262/#sec-utf16encodecodepoint))

also i think its worth mentioning that im not sure the `else` will ever actually even be reachable given the parent `if` above since i think all the (current) values of `isStrWhiteSpace`/`U16_IS_SURROGATE`/etc. are `<= 0xFFFF`

maybe im doing the bitwise math incorrectly, but if we did use `!U16_IS_SURROGATE` then wouldn't something like `RegExp.escape("0xd800")` result in something like `"\\ud7f6\\udc00"` instead of just `"\\ud800"`?

Comment (rkirsling) [reply Source/JavaScriptCore/runtime/RegExpConstructor.cpp:160]: So I was saying something rather simpler, based on the way that you've written your `if` here—either it's some specific ASCII chars, or it's `isStrWhiteSpace`, or it's `U16_IS_SURROGATE`. But I was forgetting that `isStrWhiteSpace` includes a handful of non-ASCII whitespace/line terminator characters. So ultimately I think what you've got makes sense.

File: Source/JavaScriptCore/runtime/RegExpConstructor.cpp (hunk 4 above)
Comment (rkirsling) [reply Source/JavaScriptCore/runtime/RegExpConstructor.cpp]: While it's acceptable either way based on the precedents you gave, I think it would be better to avoid `const char*`—ChrisD just landed a patch about exactly this a few hours ago: https://github.com/WebKit/WebKit/pull/27549

Comment (rkirsling) [reply Source/JavaScriptCore/runtime/RegExpConstructor.cpp]: Ah crap, I said `AtomString` when I meant `ASCIILiteral`. 🤦 I think maybe `StringView(ASCIILiteral)` would be more appropriate here, as we have in [IntlNumberFormat.cpp](https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/runtime/IntlNumberFormat.cpp#L212). Sorry for all the runaround.

**Review by rkirsling — APPROVED** [review]
